[BE4] 강찬우

loadingKKamo21 opened the pull request.

Commit:
build: 프로젝트 추가

Commit:
Merge remote-tracking branch 'origin/dev'

KangBaekGwa commented on src/main/java/org/example/Main.java:
```diff
@@ -0,0 +1,33 @@
+        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
+        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+
+        WiseSayingRepository repository =
+                LoggingProxy.createProxy(new WiseSayingRepositoryImpl(storeDir), WiseSayingRepository.class);
+        WiseSayingService service =
+                LoggingProxy.createProxy(new WiseSayingServiceImpl(repository), WiseSayingService.class);
+        WiseSayingController controller = new WiseSayingController(service, br, bw);
```
로깅을 위한 동적프록시!! b b

KangBaekGwa commented on src/main/java/org/example/controllers/WiseSayingController.java:
```diff
@@ -0,0 +1,185 @@
+            bw.write("명언: ");
+            bw.flush();
+            content = br.readLine();
+            inputEmptyCheck(content);
```
공백은 신경 안썼는데, 되게 꼼꼼하게 체킹 하셨네요... b b

KangBaekGwa commented on src/main/java/org/example/entities/BaseEntity.java:
```diff
@@ -0,0 +1,13 @@
+package org.example.entities;
+
+public abstract class BaseEntity {
+    private final int id;
+
+    protected BaseEntity(final int id) {
+        this.id = id;
+    }
+
+    public int getId() {
+        return id;
+    }
+}
```
오우.. BaseEntity 추가 좋네요!

KangBaekGwa commented on src/main/java/org/example/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,12 @@
+
+public class GlobalExceptionHandler {
+
+    public static void initialize() {
+        Thread.setDefaultUncaughtExceptionHandler((t, e) -> {
+            System.err.println("Uncaught exception in thread " + t.getName());
+            ExceptionLogger.logException((Exception) e);
+        });
+    }
```
핸들링 안된 exception을 공통 처리해서 로깅 남기는 것 또한 좋은 방법이라고 생각되네요!
요구 사항에는 없던 조항이지만, 요구사항만 지켜서 개발하는건 낭만이 아니라고 생각하고 있는데, 좋은 생각이신 것 같습니다! b

KangBaekGwa commented on src/main/java/org/example/error/ex/ErrorCode.java:
```diff
@@ -0,0 +1,19 @@
+package org.example.error.ex;
+
+public enum ErrorCode {
+
+    INVALID_ID("올바르지 않은 ID 입니다."),
+    FILE_SAVE_FAILED("파일을 저장하지 못했습니다."),
+    FILE_LOAD_FAILED("파일을 불러오지 못했습니다."),
+    FILE_DELETE_FAILED("파일을 삭제하지 못했습니다.");
+    private final String message;
+
+    ErrorCode(final String message) {
+        this.message = message;
+    }
+
+    public String getMessage() {
+        return message;
+    }
+
+}
```
ErrorCode 정의 해두는 것! b b

KangBaekGwa commented on src/main/java/org/example/repositories/WiseSayingRepositoryImpl.java:
```diff
@@ -0,0 +1,161 @@
+    private static final String DATA = "data";
+    private static final String LAST_ID = "lastId";
+
+    private final Map<Integer, WiseSaying> map;
+    private final String storeDir;
+    private int currentId;
```
저는 stateless 로 생각하고, 그냥 항상 불러오게 해놓았는데, 메모리에 올려두는게 더 성능적으로 확실히 좋아보이긴 하네요! b

loadingKKamo21 replied:
좋은 의견 감사합니다.
사실 말씀하신 것처럼 DB에서 값을 항상 가져오도록 하는 게 더 좋은 설계인 것 같습니다. 다만 값의 다양성이나 종류, 숫자(용량) 등을 고려했을 때 애플리케이션 내부 메모리를 사용하도록 했습니다. PK도 마찬가지로 생성 기준에 따라 제대로 정의하는 게 맞는 것 같지만 역시 단순처리만 했구요.
여러 모로 확장성이 크지 않은 형태입니다. 만약 좀 더 세분화하거나 기능 추가를 고려한다면 많은 부분의 수정이 필요합니다.

KangBaekGwa commented on src/test/java/org.example/AppTest.java:
제가 추구하는 통합 테스트의 방향과 가장 잘 맞는 테스트 방법인 것 같습니다! b b

KangBaekGwa reviewed (approved):
보면서 제꺼 너무 대충짠거같아서 조금 부끄러워 지네요...
안티패턴 없이, 객체지향적으로 잘 짜신거 같아서 많은 참고가 되었습니다!

oungsi2000 commented on src/main/java/org/example/dto/WiseSayingPaging.java:
저는 서비스단에 페이징 기능도 넣어놨는데 따로 빼는게 더 나은 것 같네요 !

loadingKKamo21 replied:
의견 감사합니다.
사실 저 WiseSayingPaing 클래스는 단순 DTO라서 저도 마찬가지로 서비스 로직에 페이징 기능이 있습니다.

oungsi2000 commented on src/main/java/org/example/error/ExceptionLogger.java:
예외를 따로 로깅하는 구조 좋습니다 !

kdo1999 commented on src/main/java/org/example/controllers/WiseSayingController.java:
```diff
@@ -0,0 +1,185 @@
+
+    private final WiseSayingService service;
+    private final BufferedReader br;
+    private final BufferedWriter bw;
```
저는 버퍼를 사용하는 방식은 생각하지 못하고 구현했었는데 성능적으로 이점을 가져간 거 같습니다.
하나 배워갑니다! 👍

oungsi2000 commented on src/main/java/org/example/services/WiseSayingService.java:
인터페이스로 만들어두는게 확실히 보기가 좋은 것 같아요

oungsi2000 commented on src/test/java/org.example/AppTest.java:
```diff
@@ -0,0 +1,272 @@
+        deleteAll();
+    }
+
+    private void deleteAll() {
+        try {
+            Files.walkFileTree(Paths.get(TEST_STORE_DIR), new SimpleFileVisitor<>() {
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    String fileName = file.getFileName().toString();
+                    if (fileName.endsWith(".json") || fileName.endsWith(".txt")) {
+                        Files.delete(file);
+                    }
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+        } catch (IOException e) {
+            e.printStackTrace();
+            throw new RuntimeException(e);
+        }
+    }
```
매 테스트마다 파일을 삭제하는 방법은 생각치도 못했네요..

loadingKKamo21 replied:
의견 감사합니다.
Mockito를 사용하지 않는 테스트다 보니 실제 경로 상에 파일이 저장/삭제 되는 과정을 각 테스트 단위에 영향 없이 하기 위해 넣었습니다.

oungsi2000 commented on src/test/java/org.example/AppTest.java:
```diff
@@ -0,0 +1,272 @@
+        }
+    }
+
+    private void createDummy(final int size) {
+        try {
+            for (int i = 1; i <= size; i++) {
+                WiseSayingReq req = WiseSayingReq.of("작가" + i, "명언" + i);
+                repository.saveToJson(req);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+            throw new RuntimeException(e);
+        }
+    }
```
임의로 더미 파일 생성하는 방법 저도 다음에 참고하겠습니다.

kdo1999 commented on src/main/java/org/example/controllers/WiseSayingController.java:
수정과 삭제시 저는 해당 ID의 데이터만 조회해서 유무를 판단하였는데 데이터 전체를 불러와서 해당 ID의 데이터의 유무를 판단하신 이유가 궁금합니다!
더 이점이 있는 것인지 궁금하여 질문드립니다!

loadingKKamo21 replied:
의견 감사합니다.

앱 구성 초기에 Map 객체를 컨트롤러에 담아서 사용하는 구조를 만들었을 때 작성한 코드가 남아있는 부분 같습니다. 아마 서비스 이후 과정에 대해서만 수정에 집중하다보니 놓친 부분이라 수정해야할 필요가 있는 것 같네요. 감사합니다.

말씀하신 id를 파라미터로 사용해서 수정(수정할 데이터 전송 DTO 포함)/삭제하는 과정이 더 올바른 과정 같습니다.

kdo1999 reviewed (commented):
저는 파일 확장자를 하드코딩해서 작성했는데 찬우님꺼 보면서 상수로 정의하지 않은걸 알아챈 거 같습니다😂

불변객체로 dto들 사용하신 부분들이랑 BaseEntity, 테스트등 제가 보면서 배워가는 기회가 된 거 같습니다 ㅎㅎ

예외, 로깅 등 그런 부분들에서 저는 신경을 못썼는데 보면서 되게 꼼꼼하시다는 생각이 드네요 수고하셨습니다 !

Commit:
fix: 수정/삭제 로직 수정

· Controller 불필요 부분 삭제
· 수정/삭제 로직에서 ID로 엔티티 존재 확인 코드 추가
· 수정/삭제 결과 boolean 타입 리턴으로 변경

Commit:
refactor: 수정/삭제 출력 문구 수정

Commit:
test: App 출력 문구 수정에 따른 테스트 결과 수정

Commit:
Merge remote-tracking branch 'origin/dev'

segye commented:
에러처리 부터 해서 로깅, 프록시 까지 정말 배울 점이 많았던 코드였던 것 같습니다! 잘 보고 배웠습니다!

DayDreaam commented on src/main/java/org/example/dto/WiseSayingReq.java:
DTO를 통해 컨트롤러와 서비스간 통신하는것을 이렇게 작동하는거군요??
영상으로 개념은 얼추 이해했는데 어떻게 사용하는지는 처음 알게 되었습니다.
공부가 되었습니다. 감사합니다.
